cli/src/package.rs

```diff
@@ -1,36 +1,138 @@
-
+use crate::error::CliError;
 use directories::ProjectDirs;
+use futures::future::join_all;
+use modmark_core::Resolve;
+use serde_json;
+use std::{
+    env::current_dir,
+    fs::{self, create_dir_all, File},
+    io::copy,
+    path::PathBuf,
+};
 
-pub trait Resolve {
-    fn resolve(&self, path: &str) -> Option<Vec<u8>>;
-}
+static RUNTIME: once_cell::sync::Lazy<tokio::runtime::Runtime> = once_cell::sync::Lazy::new(|| {
+    tokio::runtime::Builder::new_multi_thread()
+        .enable_all()
+        .build()
+        .unwrap()
+});
+
+pub struct PackageManager {}
 
-pub struct PackageManager;
 impl Resolve for PackageManager {
-    fn resolve(&self, path: &str) -> Option<Vec<u8>> {
-        let mut local_path = match current_dir() {
-            Ok(p) => p,
-            Err(_) => return None,
-        };
+    type Error = CliError;
+    fn resolve(&self, path: &str) -> Result<Vec<u8>, Self::Error> {
+        Ok(RUNTIME.block_on(resolve_package(path))?)
+    }
+    fn resolve_all(&self, paths: Vec<&str>) -> Vec<Result<Vec<u8>, Self::Error>> {
+        RUNTIME.block_on(resolve_packages(paths))
+    }
+}
+
+async fn resolve_packages(paths: Vec<&str>) -> Vec<Result<Vec<u8>, CliError>> {
+    let futures = paths.iter().map(|path| resolve_package(path));
+
+    return join_all(futures).await;
+}
 
-        local_path.push(PathBuf::from(path));
+async fn resolve_package(path: &str) -> Result<Vec<u8>, CliError> {
+    let splitter = path.split_once(':');
 
-        let mut cache_path = match ProjectDirs::from("org", "modmark", "packages") {
-            Some(path) => path.cache_dir().to_path_buf(),
-            None => return None,
+    let Some((specifier, package_path)) = splitter else {
+           return fetch_local(path);
         };
 
-        cache_path.push(PathBuf::from(path));
+    match specifier {
+        "http" => fetch_url(&path).await,
+        "https" => fetch_url(&path).await,
+        "pkgs" => fetch_registry(package_path).await,
+        other => Err(CliError::Specifier(other.to_string())),
+    }
+}
+
+async fn fetch_url(package_path: &str) -> Result<Vec<u8>, CliError> {
+    let mut cache_path = match ProjectDirs::from("org", "modmark", "packages") {
+        Some(path) => path.cache_dir().to_path_buf(),
+        None => return Err(CliError::Cache),
+    };
+
+    let splitter = package_path.split_once(':');
+    let Some((_, mut package_path)) = splitter else {
+        return Err(CliError::Cache)
+    };
+
+    package_path = &package_path[2..];
+
+    cache_path.push(PathBuf::from(&package_path));
+
+    let Some(path) = cache_path.parent() else {return Err(CliError::Cache)};
+
+    println!("{}", path.display());
+
+    create_dir_all(path)?;
 
-        if local_path.exists() {
-            fs::read(local_path).ok();
-        }
+    if cache_path.exists() {
+        Ok(fs::read(cache_path)?)
+    } else {
+        println!("{}", package_path);
+        let response = reqwest::get(package_path).await?;
```

Comment:
This has the wrong URL, you rewrite the package path when calculating the cache dir and because of that, we don't get `https://github.com/...`, rather `github.com/...` and this fails